Ivy

mhevery opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [n/a] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[X] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mhevery committed:
feat(core): Moving Renderer3 into @angular/core

kara reviewed (approved):
LGTM

alexeagle commented on package.json:
```diff
@@ -91,7 +92,7 @@
     "shelljs": "^0.7.8",
     "source-map": "0.5.7",
-    "source-map-support": "0.4.18",
+    "source-map-support": "^0.5.0",
```
why do you need this change?

mhevery replied:
reverted

alexeagle commented on package.json:
```diff
@@ -41,6 +41,7 @@
     "@types/node": "6.0.88",
     "@types/selenium-webdriver": "3.0.7",
     "@types/source-map": "^0.5.1",
+    "@types/source-map-support": "^0.4.0",
```
do you need this?

mhevery replied:
i do not, reverted

alexeagle commented on packages/core/src/render3/assert.ts:
```diff
@@ -0,0 +1,39 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+function stringify(value: any) {
```
why is this file in core/src/render3 ? shouldn't it be in a more common location?

mhevery replied:
we can move it later, right now just keeping it in sync with Ivy repository

alexeagle commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,184 @@
+
+  /**
+   * Which host element should the component be bootstrapped on. If not specified
+   * the component definition's `tag` is used t query the existing DOM for the
```
t -> to

mhevery replied:
fixed

alexeagle commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,184 @@
+   */
+  injector?: Injector;
+
+  features?: (<T>(component: T, componentDef: ComponentDef<T>) => void)[];
```
jsdoc?

mhevery replied:
added

alexeagle commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,184 @@
+    location: {nativeElement: getHostElement(component)},
+    injector: opts.injector || NULL_INJECTOR,
+    instance: component,
+    hostView: viewRef,
```
nit: name the variable the same as the object property so this is just `hostView` (with no key)

mhevery replied:
fixod

alexeagle commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,184 @@
+function createViewRef<T>(detectChanges: () => void, context: T): EmbeddedViewRef<T> {
+  return addDestroyable(
+      {
+        rootNodes: null !,
```
should the type of `rootNodes` accept null?

mhevery replied:
This is a hack until it is fully implemented.

jelbourn replied:
Add TODO comment that explains this?

alexeagle commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,184 @@
+          }
+        },
+        detectChanges: detectChanges,
+        checkNoChanges: function(): void {
```
why don't these use fat arrow syntax?

mhevery replied:
no reason, but not changing since I don't want to have the repos diverge

alexeagle commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,184 @@
+    destroyFn && destroyFn.forEach((fn) => fn());
+    this.destroyed = true;
+  };
+  obj.onDestroy = function(fn: Function) { (destroyFn || (destroyFn = [])).push(fn); };
```
again, function syntax (we are missing linting in this repo)

mhevery replied:
fixed

alexeagle commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,184 @@
+}
+
+
+// TODO: A hack to not pull in the NullInjector from @angular/core.
```
can you fix the hack before landing in master?

mhevery replied:
No, because it would break the other repo, will fix when we officially cut over

alexeagle commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,184 @@
+  const renderer = opts.renderer || document;
+  const componentDef = componentType.ngComponentDef;
+  let component: T;
+  const oldView = enterView(createViewState(-1, renderer), null !);
```
why doesn't the type of the last parameter admit `null`?

mhevery replied:
fixed

alexeagle commented on packages/core/src/render3/di.ts:
```diff
@@ -0,0 +1,156 @@
+
+class ElementRef implements IElementRef {
+  readonly nativeElement: any;
+  constructor(nativeElement: any) { this.nativeElement = nativeElement; }
```
why `any`?

mhevery replied:
That is how it was implemented in old angular because of WebWorkers

alexeagle commented on packages/core/src/render3/index.ts:
```diff
@@ -0,0 +1,82 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+// TODO: find out why this is needed to make closure compiler happy
```
why what is needed? the import statement?

mhevery replied:
removed

alexeagle commented on packages/core/src/render3/index.ts:
```diff
@@ -0,0 +1,82 @@
+import {ComponentDef, ComponentTemplate, ComponentType, DirectiveDef, DirectiveDefFlags, NgOnChangesFeature, PublicFeature, defineComponent, defineDirective} from './public_interfaces';
+
+// Naming scheme:
+// - Capital letters are for crating things: T(Text), E(Element), D(Directive), V(View),
```
crating -> creating

mhevery replied:
fixed

alexeagle commented on packages/core/src/render3/index.ts:
```diff
@@ -0,0 +1,82 @@
+  viewCreate as V,
+  viewEnd as v
+} from './instructions';
+// clang-format on
```
if you add a trailing comma, the formatter will put one-per-line as you want

mhevery replied:
fixed

alexeagle commented on packages/core/test/BUILD.bazel:
```diff
@@ -6,7 +6,10 @@ load("@build_bazel_rules_nodejs//:defs.bzl", "jasmine_node_test")
-    srcs = glob(["**/*.ts"]),
+    srcs = glob(
+        ["**/*.ts"],
+        exclude = ["render3/**/*.ts"],
```
This should not be needed, because glob should not descend into subpackages.
Apparently it's broken in bazel 0.8 but will be fixed in next release, according to https://github.com/bazelbuild/bazel/issues/4242

alexeagle commented on packages/core/test/render3/BUILD.bazel:
```diff
@@ -0,0 +1,27 @@
+
+jasmine_node_test(
+    name = "render3",
+    srcs = [],
```
omit `srcs`, it's no longer required

mhevery replied:
fixed

alexeagle commented on packages/platform-server/BUILD.bazel:
```diff
@@ -13,6 +13,7 @@ ts_library(
     module_name = "@angular/platform-server",
     tsconfig = "//packages:tsconfig",
     deps = [
+        "//packages:types",
```
why do you need to change platform-server?

mhevery replied:
reverted

alexeagle reviewed (commented):
I didn't review all the `.ts` code

mhevery commented:
I took care of most of these with exception of couple which would break the
original repo. I want to keep the code working in both places until we
officially cut over.


On Thu, Dec 7, 2017 at 1:50 PM, Alex Eagle <notifications@github.com> wrote:

> *@alexeagle* commented on this pull request.
>
> I didn't review all the .ts code
> ------------------------------
>
> In package.json
> <https://github.com/angular/angular/pull/20855#discussion_r155645937>:
>
> > @@ -91,7 +92,7 @@
>      "semver": "5.4.1",
>      "shelljs": "^0.7.8",
>      "source-map": "0.5.7",
> -    "source-map-support": "0.4.18",
> +    "source-map-support": "^0.5.0",
>
> why do you need this change?
> ------------------------------
>
> In package.json
> <https://github.com/angular/angular/pull/20855#discussion_r155645966>:
>
> > @@ -41,6 +41,7 @@
>      "@types/node": "6.0.88",
>      "@types/selenium-webdriver": "3.0.7",
>      "@types/source-map": "^0.5.1",
> +    "@types/source-map-support": "^0.4.0",
>
> do you need this?
> ------------------------------
>
> In packages/core/src/render3/assert.ts
> <https://github.com/angular/angular/pull/20855#discussion_r155646091>:
>
> > @@ -0,0 +1,39 @@
> +/**
> + * @license
> + * Copyright Google Inc. All Rights Reserved.
> + *
> + * Use of this source code is governed by an MIT-style license that can be
> + * found in the LICENSE file at https://angular.io/license
> + */
> +
> +function stringify(value: any) {
>
> why is this file in core/src/render3 ? shouldn't it be in a more common
> location?
> ------------------------------
>
> In packages/core/src/render3/component.ts
> <https://github.com/angular/angular/pull/20855#discussion_r155646194>:
>
> > +
> +
> +/**
> + * Options which control how the component should be bootstrapped.
> + */
> +export interface CreateComponentOptionArgs {
> +  /**
> +   * Which renderer to use.
> +   */
> +  renderer?: Renderer3;
> +
> +  rendererFactory?: RendererFactory3;
> +
> +  /**
> +   * Which host element should the component be bootstrapped on. If not specified
> +   * the component definition's `tag` is used t query the existing DOM for the
>
> t -> to
> ------------------------------
>
> In packages/core/src/render3/component.ts
> <https://github.com/angular/angular/pull/20855#discussion_r155646304>:
>
> > +
> +  rendererFactory?: RendererFactory3;
> +
> +  /**
> +   * Which host element should the component be bootstrapped on. If not specified
> +   * the component definition's `tag` is used t query the existing DOM for the
> +   * element to bootstrap.
> +   */
> +  host?: RElement|string;
> +
> +  /**
> +   * Optional Injector which is the Module Injector for the component.
> +   */
> +  injector?: Injector;
> +
> +  features?: (<T>(component: T, componentDef: ComponentDef<T>) => void)[];
>
> jsdoc?
> ------------------------------
>
> In packages/core/src/render3/component.ts
> <https://github.com/angular/angular/pull/20855#discussion_r155646497>:
>
> > +
> +/**
> + * Bootstrap a Component into an existing host element and return `ComponentRef`.
> + *
> + * @param componentType Component to bootstrap
> + * @param options Optional parameters which control bootstrapping
> + */
> +export function createComponentRef<T>(
> +    componentType: ComponentType<T>, opts: CreateComponentOptionArgs): ComponentRef<T> {
> +  const component = renderComponent(componentType, opts);
> +  const viewRef = createViewRef(detectChanges.bind(component), component);
> +  return {
> +    location: {nativeElement: getHostElement(component)},
> +    injector: opts.injector || NULL_INJECTOR,
> +    instance: component,
> +    hostView: viewRef,
>
> nit: name the variable the same as the object property so this is just
> hostView (with no key)
> ------------------------------
>
> In packages/core/src/render3/component.ts
> <https://github.com/angular/angular/pull/20855#discussion_r155646677>:
>
> > +  return {
> +    location: {nativeElement: getHostElement(component)},
> +    injector: opts.injector || NULL_INJECTOR,
> +    instance: component,
> +    hostView: viewRef,
> +    changeDetectorRef: viewRef,
> +    componentType: componentType,
> +    destroy: function() {},
> +    onDestroy: function(cb: Function): void {}
> +  };
> +}
> +
> +function createViewRef<T>(detectChanges: () => void, context: T): EmbeddedViewRef<T> {
> +  return addDestroyable(
> +      {
> +        rootNodes: null !,
>
> should the type of rootNodes accept null?
> ------------------------------
>
> In packages/core/src/render3/component.ts
> <https://github.com/angular/angular/pull/20855#discussion_r155646781>:
>
> > +  return addDestroyable(
> +      {
> +        rootNodes: null !,
> +        // inherited from core/ChangeDetectorRef
> +        markForCheck: function(): void {
> +          if (ngDevMode) {
> +            implement();
> +          }
> +        },
> +        detach: function(): void {
> +          if (ngDevMode) {
> +            implement();
> +          }
> +        },
> +        detectChanges: detectChanges,
> +        checkNoChanges: function(): void {
>
> why don't these use fat arrow syntax?
> ------------------------------
>
> In packages/core/src/render3/component.ts
> <https://github.com/angular/angular/pull/20855#discussion_r155648558>:
>
> > +  destroy(): void;
> +  onDestroy(cb: Function): void;
> +}
> +
> +function implement() {
> +  throw new Error('NotImplemented');
> +}
> +
> +function addDestroyable<T, C>(obj: any, context: C): T&DestroyRef<C> {
> +  let destroyFn: Function[]|null = null;
> +  obj.destroyed = false;
> +  obj.destroy = function() {
> +    destroyFn && destroyFn.forEach((fn) => fn());
> +    this.destroyed = true;
> +  };
> +  obj.onDestroy = function(fn: Function) { (destroyFn || (destroyFn = [])).push(fn); };
>
> again, function syntax (we are missing linting in this repo)
> ------------------------------
>
> In packages/core/src/render3/component.ts
> <https://github.com/angular/angular/pull/20855#discussion_r155648616>:
>
> > +  throw new Error('NotImplemented');
> +}
> +
> +function addDestroyable<T, C>(obj: any, context: C): T&DestroyRef<C> {
> +  let destroyFn: Function[]|null = null;
> +  obj.destroyed = false;
> +  obj.destroy = function() {
> +    destroyFn && destroyFn.forEach((fn) => fn());
> +    this.destroyed = true;
> +  };
> +  obj.onDestroy = function(fn: Function) { (destroyFn || (destroyFn = [])).push(fn); };
> +  return obj;
> +}
> +
> +
> +// TODO: A hack to not pull in the NullInjector from @angular/core.
>
> can you fix the hack before landing in master?
> ------------------------------
>
> In packages/core/src/render3/component.ts
> <https://github.com/angular/angular/pull/20855#discussion_r155648780>:
>
> > +
> +/**
> + * Bootstrap a Component into an existing host element and return `NgComponent`.
> + *
> + * NgComponent is a light weight Custom Elements inspired API for bootstrapping and
> + * interacting with bootstrapped component.
> + *
> + * @param componentType Component to bootstrap
> + * @param options Optional parameters which control bootstrapping
> + */
> +export function renderComponent<T>(
> +    componentType: ComponentType<T>, opts: CreateComponentOptionArgs = {}): T {
> +  const renderer = opts.renderer || document;
> +  const componentDef = componentType.ngComponentDef;
> +  let component: T;
> +  const oldView = enterView(createViewState(-1, renderer), null !);
>
> why doesn't the type of the last parameter admit null?
> ------------------------------
>
> In packages/core/src/render3/di.ts
> <https://github.com/angular/angular/pull/20855#discussion_r155649078>:
>
> > +        bloomBit < 64 ? (bloomBit < 32 ? di.cbf0 : di.cbf1) : (bloomBit < 96 ? di.cbf2 : di.cbf3);
> +    // Only go to parent if parent may have value otherwise exit.
> +    di = (value & mask) ? di.parent : null;
> +  }
> +  return null;
> +}
> +
> +
> +export function injectElementRef(): IElementRef {
> +  let di = getOrCreateNodeInjector();
> +  return di.elementRef || (di.elementRef = new ElementRef(di.node.native));
> +}
> +
> +class ElementRef implements IElementRef {
> +  readonly nativeElement: any;
> +  constructor(nativeElement: any) { this.nativeElement = nativeElement; }
>
> why any?
> ------------------------------
>
> In packages/core/src/render3/index.ts
> <https://github.com/angular/angular/pull/20855#discussion_r155649207>:
>
> > @@ -0,0 +1,82 @@
> +/**
> + * @license
> + * Copyright Google Inc. All Rights Reserved.
> + *
> + * Use of this source code is governed by an MIT-style license that can be
> + * found in the LICENSE file at https://angular.io/license
> + */
> +
> +// TODO: find out why this is needed to make closure compiler happy
>
> why what is needed? the import statement?
> ------------------------------
>
> In packages/core/src/render3/index.ts
> <https://github.com/angular/angular/pull/20855#discussion_r155649258>:
>
> > @@ -0,0 +1,82 @@
> +/**
> + * @license
> + * Copyright Google Inc. All Rights Reserved.
> + *
> + * Use of this source code is governed by an MIT-style license that can be
> + * found in the LICENSE file at https://angular.io/license
> + */
> +
> +// TODO: find out why this is needed to make closure compiler happy
> +import {createComponentRef, detectChanges, getHostElement, markDirty, renderComponent} from './component';
> +import {inject, injectElementRef, injectTemplateRef, injectViewContainerRef} from './di';
> +import {ComponentDef, ComponentTemplate, ComponentType, DirectiveDef, DirectiveDefFlags, NgOnChangesFeature, PublicFeature, defineComponent, defineDirective} from './public_interfaces';
> +
> +// Naming scheme:
> +// - Capital letters are for crating things: T(Text), E(Element), D(Directive), V(View),
>
> crating -> creating
> ------------------------------
>
> In packages/core/src/render3/index.ts
> <https://github.com/angular/angular/pull/20855#discussion_r155649397>:
>
> > +
> +  listenerCreate as L,
> +  memory as m,
> +  queryCreate as Q,
> +
> +  refreshComponent as r,
> +  refreshContainer as rC,
> +  refreshContainerEnd as rc,
> +  refreshQuery as rQ,
> +  textCreate as T,
> +  textCreateBound as t,
> +
> +  viewCreate as V,
> +  viewEnd as v
> +} from './instructions';
> +// clang-format on
>
> if you add a trailing comma, the formatter will put one-per-line as you
> want
> ------------------------------
>
> In packages/core/test/BUILD.bazel
> <https://github.com/angular/angular/pull/20855#discussion_r155650102>:
>
> > @@ -6,7 +6,10 @@ load("@build_bazel_rules_nodejs//:defs.bzl", "jasmine_node_test")
>  ts_library(
>      name = "test_lib",
>      testonly = 1,
> -    srcs = glob(["**/*.ts"]),
> +    srcs = glob(
> +        ["**/*.ts"],
> +        exclude = ["render3/**/*.ts"],
>
> This should not be needed, because glob should not descend into
> subpackages.
> Apparently it's broken in bazel 0.8 but will be fixed in next release,
> according to bazelbuild/bazel#4242
> <https://github.com/bazelbuild/bazel/issues/4242>
> ------------------------------
>
> In packages/core/test/render3/BUILD.bazel
> <https://github.com/angular/angular/pull/20855#discussion_r155650228>:
>
> > +    testonly = 1,
> +    srcs = glob(
> +        ["**/*.ts"],
> +        exclude = ["**/*_perf.ts"],
> +    ),
> +    tsconfig = "//packages:tsconfig.json",
> +    deps = [
> +        "//packages:types",
> +        "//packages/core",
> +        "//packages/platform-browser",
> +    ],
> +)
> +
> +jasmine_node_test(
> +    name = "render3",
> +    srcs = [],
>
> omit srcs, it's no longer required
> ------------------------------
>
> In packages/platform-server/BUILD.bazel
> <https://github.com/angular/angular/pull/20855#discussion_r155651292>:
>
> > @@ -13,6 +13,7 @@ ts_library(
>      module_name = "@angular/platform-server",
>      tsconfig = "//packages:tsconfig",
>      deps = [
> +        "//packages:types",
>
> why do you need to change platform-server?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/20855#pullrequestreview-81999776>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAG1T3wtt6UdpzQ3v-0ff9suyavpMLK4ks5s-F3BgaJpZM4Q5BIs>
> .
>

vicb commented on packages/core/src/render3/assert.ts:
```diff
@@ -0,0 +1,39 @@
+  (actual < expected) && assertThrow(actual, expected, condition, '>');
+}
+
+export function assertNotNull<T>(actual: T, condition: string) {
```
bad name

jelbourn commented on packages/core/src/render3/assert.ts:
```diff
@@ -0,0 +1,39 @@
+ */
+
+function stringify(value: any) {
+  return typeof value === 'string' ? `"${value}"` : '' + value;
```
* Add an explanation (function description) for why string get wrapped in an additional set of `"` but other types don't?
* `stringify` isn't a great name because of that extra behavior of adding the `"` (vs. only coercing the type to `string`). Could the name be more specific?

kara replied:
Yeah, maybe`stringifyWithQuotes` ?

jelbourn replied:
How about `stringifyValueForDisplay` or `stringifyValueForError`?

kara replied:
Nice, let's do `stringifyValueForError` then

jelbourn commented on packages/core/src/render3/assert.ts:
```diff
@@ -0,0 +1,39 @@
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
```
Add a file-level comment that provides some context on the functions contained within? (what they're meant to be used for)

jelbourn commented on packages/core/src/render3/assert.ts:
```diff
@@ -0,0 +1,39 @@
+  (actual == expected) && assertThrow(actual, expected, condition, '!=');
+}
+
+export function assertThrow<T>(
```
Add JsDoc description block for this? It's not immediately obvious what `condition`, `operator`, and `serializer` are from just looking at this function.

(Ideally I'd like to see description on all the functions, but I'd acknowledge the rest are mostly self-explanatory)

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+/**
+ * Options which control how the component should be bootstrapped.
+ */
+export interface CreateComponentOptionArgs {
```
Is this _always_ for bootstrapping? If so, what about `ComponentBootstrapConfig` or `ComponentBootstrapOptions`?

If it's any creation, `ComponentCreationConfig` or `ComponentCreationOptions`?

It's not immediately clear what is meant by "OptionArgs".

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+export interface CreateComponentOptionArgs {
+  /**
+   * Which renderer to use.
+   */
```
Optional: short JsDoc can be collapsed to a single line
```ts
/** Renderer instance used for the created component. */
renderer?: Renderer3;
```
This increases in information density of the file, making it easier to read more at once without scrolling.

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+   * Which host element should the component be bootstrapped on. If not specified
+   * the component definition's `tag` is used to query the existing DOM for the
+   * element to bootstrap.
+   */
```
Grammar: 
```ts
/**
 * Host element on which the component will be bootstrapped. If not specified, the component
 * definition's `tag` is used to query the existing DOM for the element to bootstrap.
 */
```

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+
+  /**
+   * Optional Injector which is the Module Injector for the component.
+   */
```
Mention what happens if left unspecified? Can also reword to be more concise
```ts
/** Module injector for the component. If unspecified, ???" */
```

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+
+  /**
+   * a set of features which should be applied to this component.
+   */
```
```ts
/** List of features applied to the created component. */
```

It would be nice to also elaborate briefly on what a "feature" is here or a `@see` for more information on features.

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+ * @param componentType Component to bootstrap
+ * @param options Optional parameters which control bootstrapping
+ */
+export function createComponentRef<T>(
```
Is this always for bootstrap? If so, would it make sense to say `bootstrapComponent<T>`? If not, `createComponent`?

This will better reflect what the function _does_ rather than what it _returns_.

jelbourn replied:
The analog I'd make is that you want APIs to be like `dialog.open()`, not `dialog.createDialogRef()`, which, while accurate, is kind of missing the point

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+export function createComponentRef<T>(
+    componentType: ComponentType<T>, opts: CreateComponentOptionArgs): ComponentRef<T> {
+  const component = renderComponent(componentType, opts);
+  const hostView = createViewRef(detectChanges.bind(component), component);
```
Use an arrow function instead of `bind`?

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+    changeDetectorRef: hostView,
+    componentType: componentType,
+    destroy: function() {},
+    onDestroy: function(cb: Function): void {}
```
Arrow functions? Can the name and type of the callback be more explicit?
```ts
destroy: () => {},
onDestroy: (callback: () => void) => {}
```

Are these not-yet-implemented? If so, add a TODO that explains this?

kara replied:
I think it's `Function` here to match the signature of the existing API. Changing to `() => void` would be slightly more restrictive right?

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+  };
+}
+
+function createViewRef<T>(detectChanges: () => void, context: T): EmbeddedViewRef<T> {
```
Does this create the view, or just the `ViewRef`? If it creates the view, call this `createView`? This will better reflect what the function _does_ rather than what it _returns_. If it's just the ref, explain that and mention where the view _does_ get created?

Add a function description? Particular, descriptions for `detectChanges` and `context` would be good because they're not perfectly obvious

kara replied:
It just creates the `ViewRef`, not the view itself. I'll add a comment.

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+  onDestroy(cb: Function): void;
+}
+
+function implement() {
```
`throwNotYetImplementedError()` or even `notYetImplemented()`?

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+      context);
+}
+
+interface DestroyRef<T> {
```
Add description for this? It's not clear to me what this is and how it would be used by just looking at the declaration

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+      {
+        rootNodes: null !,
+        // inherited from core/ChangeDetectorRef
+        markForCheck: () => {
```
Even with all of the not-yet-implemented methods, make this a class so the methods are via prototype instead of individual instances?

mhevery committed:
refactor(core): store locals in main array in rederer3

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+  throw new Error('NotImplemented');
+}
+
+function addDestroyable<T, C>(obj: any, context: C): T&DestroyRef<C> {
```
Add function description, especially for the params (`obj: any` could be anything).

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+
+// TODO: A hack to not pull in the NullInjector from @angular/core.
+export const NULL_INJECTOR: Injector = {
+  get: function(token: any, notFoundValue?: any) {
```
Could this be an arrow function?

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+
+
+/**
+ * Bootstrap a Component into an existing host element and return `NgComponent`.
```
`createComponentRef` has the same description- change this function's description to be more specific about what it does?

kara replied:
Yeah, it's a bit confusing because it `createComponentRef` wraps `renderComponent`, so technically `createComponentRef` also does what's in here. The difference is really the return value.

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+ * Bootstrap a Component into an existing host element and return `NgComponent`.
+ *
+ * NgComponent is a light weight Custom Elements inspired API for bootstrapping and
+ * interacting with bootstrapped component.
```
```
NgComponent is a lightweight API for bootstrapping and for interacting with the
bootstrapped component. The API is inspired by the native Custom Element API.
```

This functions returns `T`, not an `NgComponent`? Should there be a comment here to explain what's up with that?

kara replied:
Ah, I believe that comment is out of date. It should say that it returns an instance of the component.

jelbourn commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+  const componentDef = componentType.ngComponentDef;
+  let component: T;
+  const oldView = enterView(createViewState(-1, renderer), null);
+  try {
```
Add comment that explains which error you're targeting with this `try`? Might also be good to comment on why `try` is okay here WRT potential VM optimization issues.

jelbourn reviewed (commented):
I got just a small ways through before a meeting- I'll come back for the rest afterwards

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+  }
+}
+
+function createStaticData(
```
`createNodeStatic` or `createStaticDataForNode`?

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+ * @param value New value to write (null to remove).
+ * @param suffix Suffix to add to style's value (optional).
+ */
+export function elementStyle<T>(
```
`setElementStyle` or `updateElementStyle`?

kara replied:
Decided to keep as is

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+ * @param value Value to write. This value will be stringified.
+ *   If value is not provided than the actual creation of the text node is delayed.
+ */
+export function textCreate(index: number, value?: any): void {
```
`createTextNode`?

kara replied:
`text`

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+ * @param index Index of the node in the data array.
+ * @param value Stringified value to write.
+ */
+export function textCreateBound<T>(index: number, value: T | NO_CHANGE): void {
```
`createBoundTextNode`?

kara replied:
Discussed offline - `textBinding`

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+ * Iterates through the outputs associated with a particular event name and subscribes to
+ * each output.
+ */
+function outputCreate(outputs: (number | string)[], listener: Function): void {
```
`createOutput`?

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+ *        renaming as part of minification.
+ * @param value A value indicating if a given class should be added or removed.
+ */
+export function elementClass<T>(index: number, className: string, value: T | NO_CHANGE): void {
```
`updateElementClass`?

kara replied:
as is

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+  const rNode = typeof elementOrSelector === 'string' ?
+      ((renderer as Renderer3Fn).selectRootElement ?
+           (renderer as Renderer3Fn).selectRootElement(elementOrSelector) :
+           (renderer as Renderer3oo).querySelector !(elementOrSelector)) :
```
Is there a strong reason not to use a facade for these two types of renders throughout this file? Branching between them takes functions that would otherwise be clean and simple and makes them just complicated enough to require thinking.

kara replied:
@mhevery ?

jelbourn commented on packages/core/src/render3/renderer.ts:
```diff
@@ -0,0 +1,129 @@
+  setProperty?(name: string, value: any): void;
+}
+
+export interface RCSSStyleDeclaration {
```
I'd suggest using `RCssStyleDeclaration` even though the capital "CSS" aligns with the native type since it's more readable and is meant to abstract the native type anyway

jelbourn commented on packages/core/src/render3/renderer.ts:
```diff
@@ -0,0 +1,129 @@
+  setProperty(propertyName: string, value: string|null, priority?: string): void;
+}
+
+export interface RDOMTokenList {
```
`RDomTokenList` (same reason as above)

kara replied:
Fair

jelbourn commented on packages/core/test/render3/basic_perf.ts:
```diff
@@ -0,0 +1,74 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {C, E, T, V, c, defineComponent, e, rC, rc, v} from '../../src/render3/index';
```
Any reason to use the mini names in a test instead of the expanded names?

kara replied:
Discussed offline

jelbourn commented on packages/core/src/render3/query.ts:
```diff
@@ -0,0 +1,208 @@
+  values: any[];
+}
+
+export class QueryState_ implements QueryState {
```
What's the reasoning behind having an interface when the concrete class is going to be `QueryState_`? Why not just have the implementation as `QueryState` and skip the interface? Without knowing the rationale it feels like an unnecessary level of indirection.

jelbourn commented on packages/core/src/render3/query.ts:
```diff
@@ -0,0 +1,208 @@
+ */
+
+import {Observable} from 'rxjs/Observable';
+import {QueryList as IQueryList, Type} from '../core';
```
Add comment here explaining the alias?

jelbourn commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -0,0 +1,293 @@
+/**
```
More consistent function descriptions throughout this file?

jelbourn commented on packages/core/src/render3/di.ts:
```diff
@@ -0,0 +1,156 @@
+/**
```
Function descriptions through this file?

jelbourn commented on packages/core/src/render3/di.ts:
```diff
@@ -0,0 +1,156 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {ComponentFactory, ComponentRef as IComponentRef, ElementRef as IElementRef, EmbeddedViewRef as IEmbeddedViewRef, Injector, NgModuleRef as INgModuleRef, TemplateRef as ITemplateRef, Type, ViewContainerRef as IViewContainerRef, ViewRef as IViewRef} from '../core';
```
Add comment explaining aliases here?

jelbourn reviewed (commented):
I ran out of steam for today. Will come back for more.

marclaval committed:
feat(core): add renderer factory in render3

marclaval committed:
perf: add large table and deep tree benchmarks for render3

jelbourn commented on packages/core/src/render3/di.ts:
```diff
@@ -0,0 +1,156 @@
+}
+
+
+function notImplemented() {
```
Duplicate function (defined elsewhere as `implement`)

kara replied:
Moved to Util

jelbourn commented on packages/core/src/render3/ng_dev_mode.ts:
```diff
@@ -0,0 +1,15 @@
+  if (typeof global != 'undefined') (global as any).ngDevMode = true;
+}
+
+export const _ngDevMode = true;
```
Add a `.d.ts` file to this codebase with a declaration for `ngDevMode`? WebStorm is not happy about it without that.
```ts
declare const ngDevMode: boolean
```

kara replied:
I believe we have that in `interfaces.ts`. Is that what you mean?

```

declare global {
  const ngDevMode: boolean;
}

```

Perhaps we should move it here?

jelbourn replied:
Weird, webstorm wasn't happy with that but was fine with the `declare const`

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+ * Only `LView`s can go into `LContainer`.
+ *
+ * @param index The index of the container in the data array
+ * @param template Optional inline template
```
Missing `@param`s

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+  const native = node.native as RElement;
+
+  // In order to match current behavior, event listeners must be added for all events (including
+  // outputs).
```
As opposed to?

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+    (cleanup || (cleanup = currentView.cleanup = [])).push(eventName, native, listener, useCapture);
+  }
+
+  let mergeData: LNodeStatic|null = node.staticData !;
```
Not clear what's intended by `mergeData` here

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+  let instance;
+  const index2 = index << 1;
+  if (directive == null) {
+    // return existing
```
Elaborate? E.g., "If the directive is not specified, this is retrieving an existing directive instance".

If this function sometimes retrieves an existing instance, would a more accurate name be `getOrCreateDirective`?

jelbourn reviewed (commented):
Another batch before some meetings

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+  const directive = directives[directiveIndex << 1];
+  const oldView = enterView(hostView, element);
+  const oldNgStaticData = ngStaticData;
+  try {
```
Comment for which error is being targeted?

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+  const oldView = enterView(hostView, host);
+  try {
+    ngStaticData = template.ngStaticData || (template.ngStaticData = [] as never);
+    template(context, creationMode);
```
Feels a bit strange that "template" is invocable. Would it make sense to call this something like `renderTemplate` with the type (`ComponentTemplate`) being something like `ComponentTemplateRenderer` or `ComponentTemplateWriter`?

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+      distributedNodes[0].push(componentChild);
+    } else if (
+        (componentChild.flags & LNodeFlags.TYPE_MASK) === LNodeFlags.Element ||
+        (componentChild.flags & LNodeFlags.TYPE_MASK) === LNodeFlags.Container) {
```
Would it be worth it to create functions like `isElement`, `isContainer` to capture the bitwise operations?

kara replied:
I think this might be the only place we check element and container, weirdly. Mostly we check whether something is a view or an element, which has its own mask

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+ * @param {number} localIndex - index under which distribution of projected nodes was memorized
+ * @param {number} selectorIndex - 0 means <ng-content> without any selector
+ */
+export function contentProjection(
```
`projectContent` or `insertProjectedContent`?

kara replied:
Agreed offline on `projection`

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+
+/**
+ * Inserts previously re-distributed projected nodes. This instruction must be preceded by a call
+ * to the distributeProjectedNodes instruction.
```
Is there a programmatic way to enforce that `distributeProjectedNodes` has already been called?

kara replied:
In `valueInData`, there is an assertion to check that there are actual nodes in the memory location passed into this function.

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+
+/** The type of the NO_CHANGE constant. */
+export type NO_CHANGE = {
+  brand: 'no change detected'
```
This doesn't seem to be a real brand because it's not `private`? Couldn't you do just
```ts
export type NO_CHANGE = {
  /** Brand to ensure this type cannot match anything else */
  private _;
};
```

For the actual `NO_CHANGE` constant, you would just coerce it. This should give the same type behavior without needing the weird string literal

skreborn replied:
As far as I know, type definitions are not allowed to have `private` members.

jelbourn replied:
Ah right, you'd need to introduce a `declare`:
```ts
export declare class NO_CHANGE_TYPE {
    private _;
};

export type NO_CHANGE = typeof NO_CHANGE_TYPE;
export const NO_CHANGE = {} as NO_CHANGE;
```

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+ *
+ * @param values an array of values to diff.
+ */
+export function bindV(values: any[]): string|NO_CHANGE {
```
Would it make sense to have a pattern like
```
createInterpolationBindingWithVariableArguments
createInterpolationBindingWithOneArgument
createInterpolationBindingWithTwoArguments
...
```

kara replied:
Discussed offline

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+}
+
+/**
+ * Create an interpolation bindings with 1 arguments.
```
Add a block comment here that explains why having one function per number-of-arguments is necessary and/or beneficial?

jelbourn reviewed (commented):
Another batch before meetings

kara committed:
refactor(core): store directive defs in static data

trotyl commented on packages/core/src/render3/public_interfaces.ts:
````diff
@@ -0,0 +1,198 @@
+ *
+ * # Example
+ * ```
+ * class MyDirective {
````
Should be `MyComponent` here?

IgorMinar replied:
+1

pkozlowski-opensource committed:
feat: support queries for elements with local names

marclaval committed:
perf: add benchmarks for IV and render3 with functions (tree)

jelbourn commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+export const NG_HOST_SYMBOL = '__ngHostLNode__';
+
+/**
+ * This property gets set before entering a template.
```
As discussed offline: it would be valuable here to describe how `renderer` can take different forms, why that is, and under what circumstances it would do so.

jelbourn commented on packages/core/src/render3/renderer.ts:
```diff
@@ -0,0 +1,129 @@
+/**
+ * Functional style of API needed to create elements and text nodes.
+ */
+export interface Renderer3Fn {
```
Call this `ProceduralRenderer3`? 
See:
https://en.wikipedia.org/wiki/Functional_programming vs.
https://en.wikipedia.org/wiki/Procedural_programming

Also expand the description to explain _why_ is exists

jelbourn commented on packages/core/src/render3/renderer.ts:
```diff
@@ -0,0 +1,129 @@
+/**
+ * Object Oriented style of API needed to create elements and text nodes.
+ */
+export interface Renderer3oo {
```
Call this `ObjectOrientedRenderer3`? Expand comment to add context on _why_ it exists

jelbourn commented on packages/core/test/render3/component_spec.ts:
```diff
@@ -0,0 +1,62 @@
+  beforeEach(
+      () => {
+
+      });
```
Empty `beforeEach`

kara commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+  ngDevMode && assertNotNull(component, 'component');
+  if (!isDirty) {
+    isDirty = true;
+    scheduler(detectChanges.bind(null, component));
```
Chatted with Misko about this and there appear to be a few tradeoffs: 

- bind is probably slower because it's megamorphic at the call site (always different shapes of functions being passed in)
- bind saves a bit on bytes because arrow functions transpile back into `function() {}`, which is a little longer

I think here switching over to arrow functions make sense

marclaval committed:
perf: use style.setProperty instead of setAttribute in render3 and iv benchmarks

kara commented on packages/core/src/render3/interfaces.ts:
```diff
@@ -0,0 +1,624 @@
+   * We use `bindingIndex` to internally keep track of which binding
+   * is currently active.
+   *
+   * NOTE: We also use nodes == null as a marker for creationMode. We
```
Yeah, this should be `data == null`.  We changed the name because `nodes` is no longer accurate -- this array  now contains local variables, nodes, and directive instances.

kara committed:
docs(core): add comments to assert functions

kara committed:
docs(core): add comments to node_manipulation functions

kara commented on packages/core/src/render3/di.ts:
```diff
@@ -0,0 +1,156 @@
+  let di = getOrCreateNodeInjector();
+  const data = (di.node as LContainer).data;
+  if (data === null || data.template === null) {
+    throw createError('Directive not used in structural way.', null);
```
"Directive does not have a template" ?

kara committed:
docs(core): document di and minor renames

kara committed:
docs(core): add more comments to di and fix formatting

kara commented on packages/core/src/render3/interfaces.ts:
```diff
@@ -0,0 +1,624 @@
+/**
+ * NOTES:
+ *
+ * Each Array costs 70 bytes and is composed of `Array` and `(array)` object
```
I think it's a general note to remind us of the cost when choosing between objects and arrays in a design. Where do you think a better place would be for it to go?

kara commented on packages/core/src/render3/interfaces.ts:
```diff
@@ -0,0 +1,624 @@
+ * don't have to edit the nodes array or directives array based on which views
+ * are present.
+ */
+export interface ViewState {
```
`LNode` is to differentiate between `LNode` (our representation of a DOM node) and `RNode` (the actual rendered DOM node), but the `ViewState` doesn't have a renderer equivalent.  I think the addition of the `L` would imply the existence of `RViewState` ? @mhevery  thoughts?

kara replied:
`LNode` is to differentiate between `LNode` (our representation of a DOM node) and `RNode` (the actual rendered DOM node), but the `ViewState` doesn't have a renderer equivalent.  I think the addition of the `L` would imply the existence of `RViewState`, which might be confusing ? @mhevery  thoughts?

kara commented on packages/core/src/render3/interfaces.ts:
```diff
@@ -0,0 +1,624 @@
+/**
+ * Abstract node container which contains other views.
+ */
+export interface LContainer extends LNode {
```
@mhevery Do you have an opinion on this?

kara commented on packages/core/src/render3/interfaces.ts:
```diff
@@ -0,0 +1,624 @@
+  next: LNode|null;
+
+  /**
+   * If ViewState, then `data` contains lightDOM.
```
This comment looks outdated, I'll updated it

kara commented on packages/core/src/render3/interfaces.ts:
```diff
@@ -0,0 +1,624 @@
+  readonly view: ViewState;
+
+  /** The injector associated with this node. Necessary for DI. */
+  nodeInjector: LNodeInjector|null;
```
I think this is to differentiate between an `Injector` and an `LNodeInjector` (which contains both the bloom filters and an `Injector`).

kara commented on packages/core/src/render3/interfaces.ts:
```diff
@@ -0,0 +1,624 @@
+   * The next sibling node. Necessary so we can propagate through the root nodes of a view
+   * to insert them or remove them from the DOM.
+   */
+  next: LNode|null;
```
@mhevery do you have a preference here? I can see it both ways

kara commented on packages/core/src/render3/interfaces.ts:
```diff
@@ -0,0 +1,624 @@
+ *
+ * In order to be fast, all subtypes of `LNode` should have the same shape.
+ * Because size of the `LNode` matters, many fields have multiple roles depending
+ * on the `LNode` subtype.
```
Maybe we could link here: https://en.wikipedia.org/wiki/Inline_caching#Monomorphic_inline_caching

kara commented on packages/core/src/render3/interfaces.ts:
```diff
@@ -0,0 +1,624 @@
+  /**
+   * The following bloom filter determines whether a directive is available
+   * on the associated node or not. This prevents us from searching the directives
+   * array at this level unless it's probable the directive is in it.
```
Maybe https://en.wikipedia.org/wiki/Bloom_filter?

kara commented on packages/core/src/render3/interfaces.ts:
```diff
@@ -0,0 +1,624 @@
+   * When a view is destroyed, listeners need to be released
+   * and onDestroy callbacks need to be called. This cleanup array
+   * stores both listener data (in chunks of 4) and onDestroy data
+   * (in chunks of 2), as they'll be processed at the same time.
```
Technically, what we do is go through all the listeners and all the onDestroys in a given *view* at the same time.  It's not so much because we care that they happen in that order, but because it saves us having to have two for loops per destroyed view (one for listeners and one for onDestroys) instead of one. I'll update the comment to clarify

kara committed:
refactor(core): renamed and split out interfaces

kara commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+/////////////
+
+
+export const refreshComponent:
```
Discussed offline -> `componentRefresh`

mhevery committed:
style: fix formatting errors

mhevery committed:
style: fix formatting errors

kara commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+ *
+ * @param index The index of the container in the data array
+ */
+export function refreshContainer(index: number): void {
```
Discussed offline - `containerRefreshStart`

kara commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+  bloomAdd(getOrCreateNodeInjector(), def.type);
+}
+
+export function directiveLifeCycle(
```
It's used in the lifecycle_spec, but it definitely needs a comment

kara committed:
docs(core): add docs to instructions, minor renames

kara committed:
refactor(core): rename instructions for consistency

kara commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+ * @param value New value to write.
+ */
+
+export function elementProperty<T>(index: number, propName: string, value: T | NO_CHANGE): void {
```
as is

kara commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+ *        renaming as port of minification.
+ * @param value Value to write. This value will go through stringification.
+ */
+export function elementAttribute(index: number, attrName: string, value: any): void {
```
as is

IgorMinar replied:
ouch 😇

kara commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+ * @param listener The function to be called when event emits
+ * @param useCapture Whether or not to use capture in event listener.
+ */
+export function listenerCreate(
```
`listener`

kara commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1590 @@
+/**
```
Yeah, the problem is the state variables. Most of the functions use them in some way, and the ones that don't are pretty small and focused, so I'm not sure they deserve their own file. We do have a few files like `di.ts` and `query.ts` that are separate - we should try to develop more new features this way.

kara commented on packages/core/src/render3/component.ts:
```diff
@@ -0,0 +1,187 @@
+}
+
+export function detectChanges<T>(component: T) {
+  ngDevMode && assertNotNull(component, 'component');
```
@mhevery What do you think?

IgorMinar replied:
I believe the plan is to strip the code from production bundles.

@mhevery @pkozlowski-opensource can you confirm?

pkozlowski-opensource replied:
@jelbourn - @IgorMinar is right - you are seeing this `ngDevMode && doSth()` pattern so we can tree-shake `doSth()` in a prod mode. We could do `isDevMode() && doSth()` but it would still require a check at each call site.

kara committed:
docs(core): add missing docs to component and fix formatting

pkozlowski-opensource committed:
refactor: remove import circular dependencies

This PR fixes a circular dependency among those files in Renderer3:
`query` -> `di` -> `instructions` -> `query` -> ...

Looking at the above dependencies the `di` -> `instructions` import is
a problematic one. Previously `di` had an import from `instructions`
since we can known about "current node" only in `instructions`
(and we need "current node" to create node injector instances).

This commit refactors the code in the way that functions in the
`di` file don't depend on any info stored module-global variables
in `instructions`.

marclaval committed:
test(core): animation renderer factory in render3

jelbourn reviewed (approved):
LGTM

IgorMinar commented on modules/benchmarks/e2e_test/largetable_perf.ts:
```diff
@@ -56,6 +56,24 @@ describe('largetable benchmark perf', () => {
+        }).then(done, done.fail);
+      });
+
+      it('should run for iv', (done) => {
```
ivy

IgorMinar commented on modules/benchmarks/e2e_test/tree_data.ts:
```diff
@@ -49,6 +49,24 @@ export const Benchmarks: Benchmark[] = [
+    ignoreBrowserSynchronization: true,
+  },
+  {
+    id: `deepTree.iv`,
```
ivy

IgorMinar commented on modules/benchmarks/e2e_test/tree_data.ts:
```diff
@@ -49,6 +49,24 @@ export const Benchmarks: Benchmark[] = [
+  },
+  {
+    id: `deepTree.iv`,
+    url: 'all/benchmarks/src/tree/iv/index.html',
```
rename the dir to ivy

IgorMinar commented on modules/benchmarks/src/largetable/iv/index.html:
```diff
@@ -0,0 +1,31 @@
+<!doctype html>
```
rename directory to ivy

IgorMinar commented on packages/core/src/render3/index.ts:
```diff
@@ -0,0 +1,84 @@
+  elementClass as k,
+  elementEnd as e,
+  elementProperty as p,
+  elementStart as E,
```
since you turned off clang-format you might as well group elementStart and elementEnd together :)

IgorMinar reviewed (commented):
incomplete review - posting comments I have so far

IgorMinar commented on packages/core/test/render3/basic_perf.ts:
```diff
@@ -0,0 +1,74 @@
+        const start = new Date().getTime();
+        renderComponent(Component);
+        const end = new Date().getTime();
+        log(`${count} DIVs in Render3`, (end - start) / count);
```
what's this test about? there no assertion of any kind. I assume it's just for manual testing and exploration.

can you please document this purpose in the file overview and move the test into some descriptive directory, e.g. 
 `test/manual/` or something like that?

IgorMinar reviewed (commented):
todo:
- [ ] make ci green

to consider:
- [ ] reformat all code using the latest clang-format from #20884

followup tasks:
- [ ] wire up the benchmarks in g3

robwormald commented on packages/core/src/render3/instructions.ts:
```diff
@@ -0,0 +1,1740 @@
+/**
+ * State of the current view being processed.
+ */
+let currentView: ViewState = createViewState(null !, null !, []);
```
this (apparently!) is invalid ES6 - it throws when current view is `let currentView = ` as opposed to `var currentView = ` 
see https://jsfiddle.net/au4y6u4b/1/

pkozlowski-opensource committed:
feat(core): support 'read' option for ngIvy queries

mhevery committed:
build: yarn buildifier

mhevery committed:
build: fix benchmarks for render3

alxhub commented on packages/core/src/render3/l_node.ts:
```diff
@@ -0,0 +1,270 @@
+
+import {ElementRef, Injector, TemplateRef, ViewContainerRef} from '../core';
+
+import {ContainerState, ProjectionState, QueryState, ViewState} from './interfaces';
```
This introduces a cyclic dependency between `interfaces.ts` and `l_node.ts` which breaks Closure compilation in g3, even though Typescript is happy to deal with that.

mhevery committed:
build: fix circular dep between interface and l_node by merging
